File: algorithms/code/hackerrank/extra_long_factorials/extra_long_factorials.rs
// https://www.hackerrank.com/challenges/extra-long-factorials

use crate::hackerrank::tester::{read_input, write_and_check_output};
use num_bigint::BigInt;

pub fn extra_long_factorials(n: u32) -> String {
    let mut result = BigInt::from(1);
    for i in 2..=n {
        result *= BigInt::from(i);
    }
    result.to_string()
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let mut input = read_input(origin, name);
    let mut results: Vec<String> = Vec::new();
    for line in &mut input {
        let n = line.trim().parse::<u32>().unwrap();
        results.push(extra_long_factorials(n));
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("_example");
    }
}

File: algorithms/code/hackerrank/extra_long_factorials/extra_long_factorials.go
// Package extralongfactorials implements https://www.hackerrank.com/challenges/extra-long-factorials
package extralongfactorials

import "math/big"

// ExtraLongFactorials - implements the solution to the problem
func ExtraLongFactorials(n int32) string {
	result := new(big.Int).SetInt64(1)
	for i := 2; int32(i) <= n; i++ {
		result.Mul(result, new(big.Int).SetInt64(int64(i)))
	}
	return result.String()
}
